src/config/system.rs

```diff
@@ -6,13 +6,213 @@ pub const MAGIC: u32 = 0xb153;
+
+#[derive(Debug)]
+pub struct RegistryChannel {
+    registry: Lazy<StorageRegistry>,
+    notify: Notify,
+}
+
+//WIP for a session typed pattern
+// See: https://stanford-cs242.github.io/f19/lectures/09-1-session-types.html
+// pub trait HasDualMetrics {
+//     type DualMetrics;
+// }
+// impl HasDualMetrics for MetricSender {
+//     type DualMetrics = MetricReceiver;
+// }
+
+// impl HasDualMetrics for MetricReceiver {
+//     type DualMetrics = MetricSender;
+// }
+// Pub struct DualMetricsChannel <T: HasDualMetrics> {}
+// impl DualMetricsChannel<MetricSender> {}
+// impl DualMetricsChannel<MetricReceiver> {}
+
+// #[cfg(feature = "prometheusd")]
+#[derive(MetricStorage, Clone, Debug)]
+#[metric(subsystem = "rpc")]
+pub struct RpcMetrics {
+    #[metric(labels("path", "method", "status"), help = "Total number of requests")]
+    requests: prometheus::IntCounterVec,
+    #[metric(labels("path", "method"), help = "latency of rpc calls")]
+    duration: prometheus::HistogramVec,
+}
+// #[cfg(feature = "prometheusd")]
+impl RpcMetrics {
+    pub fn init(registry: &StorageRegistry) -> Result<&Self, prometheus::Error> {
+        RpcMetrics::instance(registry)
+    }
+    pub fn requests_complete(&self, path: &str, method: &str, status: &u16, duration: Duration) {
+        let dt = duration.as_millis() as f64;
+        self.requests
+            .with_label_values(&[path, method, &status.to_string()])
+            .inc();
+        self.duration.with_label_values(&[path, method]).observe(dt)
+    }
+    pub fn push_latency(&self, path: &str, method: &str, dt: f64) {
+        self.duration.with_label_values(&[path, method]).observe(dt)
+    }
+}
 
 #[cfg(feature = "journald")]
 pub fn log_journald(level: u32, message: &str) {
     use systemd::journal;
     journal::print(level, message);
 }
 
+// #[cfg(feature = "prometheusd")]
+pub struct RpcMetricsReciever {
+    inner: MetricReceiver,
+    pub name: &'static str,
+    metrics: Option<RpcMetrics>,
+}
+// #[cfg(feature = "prometheusd")]
+impl std::fmt::Debug for RpcMetricsReciever {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("RpcMetricsReciever").finish()
+    }
+}
+// #[cfg(feature = "prometheusd")]
+pub struct RpcMetricsSender {
+    inner: MetricSender,
+    pub name: &'static str,
+    metrics: Option<RpcMetrics>,
+}
+// #[cfg(feature = "prometheusd")]
+impl std::fmt::Debug for RpcMetricsSender {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("RpcMetricsSender").finish()
+    }
+}
+
+#[derive(Debug)]
+enum MetricsCommand<'a> { 
+    Flush(&'a RegistryChannel),
+    Channel(&'a RpcMetricsSender, &'a RpcMetricsReciever),
+    Push(&'a RpcMetrics, &'a RegistryChannel),
+}
+
+#[derive(Debug)]
+enum MetricsSenderCommand {
+    AdminMsg(oneshot::Sender<RpcMetrics>),
+    StatsMsg(oneshot::Sender<RpcMetrics>),
+}
+
+// #[cfg(feature = "prometheusd")]
+impl RegistryChannel {
```

Comment:
this is probably overcomplicating it. I imagined it looking something like the new liveness updater: https://github.com/rainshowerLabs/blutgang/blob/439b7186cdc235ebb2e01a6ec8bc2dcbb9a536c4/src/admin/listener.rs#L127

https://github.com/rainshowerLabs/blutgang/blob/439b7186cdc235ebb2e01a6ec8bc2dcbb9a536c4/src/admin/accept.rs#L141

The idea is to spawn it as a seperate process and just clone the mpsc sender to every ws connection, http request as needed. we would keep the ownership of the prometheus data within that specific thread and could interact with the admin rpc to send the stats when queried much easier.

so what id do, is id spawn the receiving end in as part of the admin api startup sequence, with a sink in case its disabled(see: https://github.com/rainshowerLabs/blutgang/blob/439b7186cdc235ebb2e01a6ec8bc2dcbb9a536c4/src/main.rs#L158). and then make 2 clones of the receiver, one sent for WS and one for HTTP. I would then clone them as a part of their respective `accept` macros along with all the other arguments needed. Inside of the method i would use them to send the raw data, to be processed async inside of the prometheus metrics thread. This is important because the timeloss from doing metrics is now equivalent to sending data through a channel as all the processing is done in a separate process.

lmk your thoughts, i am not familiar with the semantics of the `prometheus` crate so it might be incompatible with how im imagining things.